v2/tools/generator/internal/codegen/pipeline/assemble_oneof.go

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+	"github.com/pkg/errors"
+)
+
+const AssembleOneOfTypesID = "assembleOneOfTypes"
+
+func AssembleOneOfTypes() *Stage {
+	stage := NewStage(
+		AssembleOneOfTypesID,
+		"Assemble OneOf types from OpenAPI Fragments",
+		func(ctx context.Context, state *State) (*State, error) {
+			assembler := newOneOfAssembler(state.Definitions())
+
+			newDefs := assembler.assembleOneOfs()
+			return state.WithDefinitions(
+					state.Definitions().OverlayWith(newDefs)),
+				nil
+		})
+
+	return stage
+}
+
+type oneOfAssembler struct {
+	// A set of all the TypeDefinitions we're assembling from
+	defs astmodel.TypeDefinitionSet
+	// A map of all the leaf OneOfs, indexed by the root they reference
+	leavesByRoot map[astmodel.TypeName][]astmodel.TypeName
+	// The set of definitions we've modified
+	updatedDefs astmodel.TypeDefinitionSet
+}
+
+// newOneOfAssembler creates a new assembler to build our oneOf types
+func newOneOfAssembler(defs astmodel.TypeDefinitionSet) *oneOfAssembler {
+	result := &oneOfAssembler{
+		defs:        defs,
+		updatedDefs: make(astmodel.TypeDefinitionSet, 100),
+	}
+
+	result.indexLeaves()
+
+	return result
+}
+
+// assembleOneOfs is called to build up the oneOf types
```

Comment:
maybe more here to explain about assembling into `what` structure?

Reply:
I'll add some docs, but not here. There's a bunch to explain.